### PR TITLE
Android Stylus pressure and tilt support.

### PR DESCRIPTION
**Pull request by Distantz**
Modified the Android input code to send pressure and tilt information.

*Bugsquad edit:*
- Implement Android part of https://github.com/godotengine/godot-proposals/issues/735

### CODE REVIEW

**Comment by Distantz** [comment]
Have verified that this works on my Galaxy Note 7.

File: platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java [hunk 1]
```diff
@@ -465,25 +465,44 @@ static boolean handleMotionEvent(int eventSource, int eventAction, int buttonsMa
...
+
+		// orientation is returned as a radian value between 0 to pi clockwise or 0 to -pi counterclockwise
+		final float orientation = event.getOrientation();
+
+		// tilt is zero is perpendicular to the screen and pi/2 is flat on the surface.
+		final float tilt = event.getAxisValue(MotionEvent.AXIS_TILT);
+
+		float tiltMult = (float)Math.sin(tilt);
+
+		// to be consistent with expected tilt
```
Comment (akien-mga) [inline platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java:484]:
```suggestion
		// Orientation is returned as a radian value between 0 to pi clockwise or 0 to -pi counterclockwise.
		final float orientation = event.getOrientation();

		// Tilt is zero is perpendicular to the screen and pi/2 is flat on the surface.
		final float tilt = event.getAxisValue(MotionEvent.AXIS_TILT);

		float tiltMult = (float)Math.sin(tilt);

		// To be consistent with expected tilt.
```

File: platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java [hunk 2]
```diff
@@ -465,25 +465,44 @@ static boolean handleMotionEvent(int eventSource, int eventAction, int buttonsMa
 	}
 
 	static boolean handleMouseEvent(final MotionEvent event) {
+
 		final int eventAction = event.getActionMasked();
```
Comment (akien-mga) [inline platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java:469]: Unnecessary empty line.
```suggestion
	static boolean handleMouseEvent(final MotionEvent event) {
		final int eventAction = event.getActionMasked();
```

File: platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java [hunk 3]
```diff
@@ -465,25 +465,44 @@ static boolean handleMotionEvent(int eventSource, int eventAction, int buttonsMa
...
-		return handleMouseEvent(eventAction, buttonsMask, x, y, horizontalFactor, verticalFactor, false, sourceMouseRelative);
+		return handleMouseEvent(eventAction, buttonsMask, x, y, horizontalFactor, verticalFactor, false, sourceMouseRelative, pressure, tiltX, tiltY);
 	}
 
 	static boolean handleMouseEvent(int eventAction, int buttonsMask, float x, float y) {
 		return handleMouseEvent(eventAction, buttonsMask, x, y, 0, 0, false, false);
 	}
 
-	static boolean handleMouseEvent(int eventAction, int buttonsMask, float x, float y, float deltaX, float deltaY, boolean doubleClick, boolean sourceMouseRelative) {
+	static boolean handleMouseEvent(int eventAction, int buttonsMask, float x, float y, float deltaX, float deltaY, boolean doubleClick, boolean sourceMouseRelative){
```
Comment (akien-mga) [inline platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java:501]:
```suggestion
	static boolean handleMouseEvent(int eventAction, int buttonsMask, float x, float y, float deltaX, float deltaY, boolean doubleClick, boolean sourceMouseRelative) {
```

**Comment by akien-mga** [comment]
Thanks for the contribution! Did you check if this solves any existing bug report or feature proposal?

Please make changes to the PR by _amending_ the commit and _force pushing_ the branch, so style fixes for example become part of the main commit, instead of follow up fixup commits. See [PR workflow](https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html#modifying-a-pull-request) for details.

File: platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java [hunk 4]
```diff
@@ -503,6 +522,12 @@ static boolean handleMouseEvent(int eventAction, int buttonsMask, float x, float
 		// follow ACTION_DOWN and ACTION_UP events. As such, handling them would result in duplicate
 		// stream of events to the engine.
 		switch (eventAction) {
+
+			// s pen weirdness
+			//case 211:
+			//case 212:
+			//case 213:
+			//case 214:
```
[the remark this answers is not in the document]
Comment (Distantz) [reply platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java:530]: Sorry, was experimenting with button codes but it isn't relevant to functionality.

**Comment by Distantz** [comment]
I believe it satisfies the Android half of godotengine/godot-proposals#735.

**Comment by akien-mga** [comment]
> I believe it satisfies the Android half of [godotengine/godot-proposals#735](https://github.com/godotengine/godot-proposals/issues/735).

Thanks! Please also have a look at #49609 for comparison - it looks like your PR is more feature complete as it also implements tilt and not just pressure. The discussion in the PR might still be useful to make this one fully ready to merge.

CC @HEAVYPOLY

**Comment by Distantz** [comment]
Had a look at #49609, seem's like they had a couple of issues with getting the stylus to register up and down events. From what I can see testing right now, my PR does emit these events.

**Comment by Distantz** [comment]
image
Logcat output from my tablet for a stylus press, move and release.

**Comment by akien-mga** [comment]
> Have verified that this works on my Galaxy Tab S7.

Did some testing to get another data point, before remembering that my device is also a Galaxy Tab S7 :P 

Anyway, I can confirm it seems to work well on that device, with a generic stylus I got with it.

Screenshot_20230815_110059_Godot Editor 4 (debug)

Screenshot of received events while writing my name with the stylus.

**Comment by Distantz** [comment]
Good to know its not a fluke haha

**Review by akien-mga — APPROVED** [review]
Tested functionality, works fine for a basic test (didn't try it on a real app that would make use of pressure and tilt). Code style looks good.

I'll let @godotengine/android review the actual implementation for Android.

**Comment by Calinou** [comment]
Do we have a testing project that features pen drawing somewhere?

**Comment by Distantz** [comment]
I've got a project I've been using to test this that displays the pressure (in both text and colour background) alongside the tilt, can post that here if needed.

**Comment by Calinou** [comment]
> I've got a project I've been using to test this that displays the pressure (in both text and colour background) alongside the tilt, can post that here if needed.

Please upload it here :slightly_smiling_face:

**Comment by Distantz** [comment]
[PressureTest.zip](https://github.com/godotengine/godot/files/12343415/PressureTest.zip)

Here you are. Hopefully it isn't too sloppy code-wise.

**Comment by ModProg** [comment]
I'll probably not be able to experiment with this before the beginning of September, unless someone builds the APK for me 😬.

**Comment by Distantz** [comment]
@ModProg I am able to give you the APK, but GitHub only allows max 25mb as attachments. Any way to send it to you?

**Comment by Zireael07** [comment]
Won't a zipped apk fit under the limit?

**Comment by Distantz** [comment]
Mine is 37mb at max compression, although there might be a way to shrink it in Godot that I don't know about?

**Comment by Calinou** [comment]
> Mine is 37mb at max compression (for reference, the raw APK is 40mb), although there might be a way to shrink it in Godot that I don't know about?

It can be shrunk by only including a single architecture (arm64 nowadays is ubiquitous on phones and tablets).

Here's the release APK I built (arm64 only): 
[android_release.apk.zip](https://github.com/godotengine/godot/files/12343921/android_release.apk.zip)

**Comment by Distantz** [comment]
Legendary, @ModProg this is for you 👍

**Comment by Calinou** [comment]
I've tested the project locally and it works as expected (Samsung Galaxy Z Fold4, S Pen Fold Edition).

Note that if the pen appears to do nothing, make sure it's enabled in the system settings.

**Review by Calinou — APPROVED** [review]

**Comment by ModProg** [comment]
I think I ran into a similar problem I had when implementing this IIRC.

There is only an output when I start moving, simply pressing down doesn't do anything independent from how hard I press down.

https://github.com/godotengine/godot/assets/11978847/73a1fe50-6b63-44c7-a098-a0943a3d5fdc

**Comment by ModProg** [comment]
Slow movements also don't trigger it. Thou I haven't looked at the Testprojekt, might be it is written to behave this way?

**Comment by Distantz** [comment]
Only InputEventMouseMotion has a pressure field, so the InputEventMouseButton events don't have any pressure information. I think there must be some sort of threshold built-in for that. The test project I supplied only displays motion events

**Comment by ModProg** [comment]
One thing I noticed in my testing, was that pen touching wasn't detected at all, because Samsung introduced custom actions: https://github.com/ModProg/godot/commit/cbce6a1bc26ef3dcffc6156cac7e5b67ff4a2b81

**Comment by Distantz** [comment]
I think supporting the stylus buttons is a different issue and should probably be in its own PR. If we do support the Samsung codes, could that potentially cause issues relating to duplicate events?

**Comment by ModProg** [comment]
> I think supporting the stylus buttons is a different issue and should probably be in its own PR. If we do support the Samsung codes, could that potentially cause issues relating to duplicate events? 

IIRC they just didn't send the nornal codes. But it would also be interesting whether e.g. Huawei sends the same or different codes. I don't think I ever researched that.

**Comment by ModProg** [comment]
But I agree, this is not necessarily a blocker for this PR. When I hopefully continue the project I wanted pen support for, I'll revisit this and can add those then.

**Comment by Distantz** [comment]
I see, thank you for the clarification!

File: platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java [hunk 5]
```diff
@@ -433,7 +433,7 @@ static boolean isMouseEvent(MotionEvent event) {
 	}
 
 	private static boolean isMouseEvent(int eventSource) {
-		boolean mouseSource = ((eventSource & InputDevice.SOURCE_MOUSE) == InputDevice.SOURCE_MOUSE) || ((eventSource & (InputDevice.SOURCE_TOUCHSCREEN | InputDevice.SOURCE_STYLUS)) == InputDevice.SOURCE_STYLUS);
```
Comment (m4gr3d) [inline platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java:436]: What's the reason for removing the `SOURCE_TOUCHSCREEN` check?

**Comment by m4gr3d** [comment]
Asides from the comment above the logic looks good!

> Only InputEventMouseMotion has a pressure field, so the InputEventMouseButton events don't have any pressure information.

`InputEventMouseMotion` extends from `InputEventMouseButton`, so it's worth considering whether we should move the `tilt` and `pressure` fields into `InputEventMouseButton` instead.

@Distantz Thoughts?

File: platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java (hunk 5 above)
Comment (Distantz) [reply platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java:436]: I couldn't get it to work without removing that. I think `SOURCE_TOUCHSCREEN` is used for fingers also?

**Comment by Distantz** [comment]
I think that could definitely be a good change, but I think it would require some conversation with the other platforms. For Android specifically, it should be easy to implement.

**Comment by akien-mga** [comment]
> Asides from the comment above the logic looks good!
> 
> > Only InputEventMouseMotion has a pressure field, so the InputEventMouseButton events don't have any pressure information.
> 
> `InputEventMouseMotion` extends from `InputEventMouseButton`, so it's worth considering whether we should move the `tilt` and `pressure` fields into `InputEventMouseButton` instead.

To be clear, the inheritance tree is like this:

- `InputEventMouse`
  * `InputEventMouseButton`
  * `InputEventMouseMotion`

So indeed the `tilt` and `pressure` properties could be made available to both `Button` and `Motion` events if moved to the base `InputEventMouse`.

This is IMO outside the scope of this PR though, and we'd need to evaluate whether moving properties to a parent class like this will break API or ABI compatibility.

File: platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java (hunk 5 above)
Comment (akien-mga) [reply platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java:436]: Should we try it like this, or investigate further?

Comment (Distantz) [reply platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java:436]: From what I can test, there seems to be no regression in touch handling when removing that `SOURCE_TOUCHSCREEN`, I think it should be fine to try it like this.

**Comment by akien-mga** [comment]
Let's proceed and merge, if @m4gr3d still has concerns on the change to `isMouseEvent`, we can improve it later on.

**Comment by akien-mga** [comment]
Thanks! And congrats for your first merged Godot contribution :tada:

File: platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java (hunk 5 above)
Comment (Sauermann) [reply platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java:436]: It looks like this change has introduced the following regression: #88621

**Pull request merged**